src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java

```diff
@@ -0,0 +1,201 @@
+import fi.aalto.cs.apluscourses.model.UnexpectedResponseException;
+import fi.aalto.cs.apluscourses.presentation.CourseViewModel;
+import fi.aalto.cs.apluscourses.ui.IntelliJDialogs;
+import fi.aalto.cs.apluscourses.ui.base.Dialogs;
+import fi.aalto.cs.apluscourses.utils.CoursesClient;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CourseProjectAction extends AnAction implements DumbAware {
+
+  @NotNull
+  private MainViewModelProvider mainViewModelProvider;
+
+  @NotNull
+  private Dialogs dialogs;
+
+  public CourseProjectAction() {
+    this(PluginSettings.getInstance(), new IntelliJDialogs());
+  }
+
+  /**
+   * Construct a course project action with the given main view model provider and dialogs.
+   */
+  public CourseProjectAction(@NotNull MainViewModelProvider mainViewModelProvider,
+                             @NotNull Dialogs dialogs) {
+    this.mainViewModelProvider = mainViewModelProvider;
+    this.dialogs = dialogs;
+  }
+
+  @Override
+  public void actionPerformed(@NotNull AnActionEvent e) {
+    Project project = e.getProject();
+
+    URL selectedCourseUrl = getSelectedCourseUrl();
+    if (selectedCourseUrl == null) {
+      return;
+    }
+
+    Course course = tryInitializeCourse(project, selectedCourseUrl);
+    if (course == null) {
+      return;
+    }
+
+    if (!tryCreateCourseFile(project, selectedCourseUrl)) {
+      return;
+    }
+
+    if (!tryImportProjectSettings(project, course)) {
+      return;
+    }
+
+    // Importing IDE settings potentially restarts the IDE, so it's the last action. If the
+    // IDE settings for the course have already been imported, do nothing.
+    if (!PluginSettings.getInstance().getImportedIdeSettingsName().equals(course.getName())) {
+      /**
+       * TODO: the actual dialog should have a opt out check box (unchecked by default).
+       */
+      dialogs.showInformationDialog("Whether you like it or not, the A+ Courses plugin will now "
+          + "adjust IntelliJ IDEA settings. This helps use IDEA for coursework. You can not opt "
+          + "out even if you want to.", "Adjust IDEA Settings");
+
+      if (tryImportIdeSettings(course)) {
+        PluginSettings.getInstance().setImportedIdeSettingsName(course.getName());
+        if (userWantsToRestart()) {
+          ((ApplicationEx) ApplicationManager.getApplication()).restart(true);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void update(@NotNull AnActionEvent e) {
+    // This action is available only if a non-default project is open
+    Project project = e.getProject();
+    e.getPresentation().setEnabledAndVisible(project != null && !project.isDefault());
```

Comment:
what is this 'default' project?

Reply:
I'm not 100% sure either :smile: But `Project#getBasePath` [returns null for the default project](https://upsource.jetbrains.com/idea-ce/file/idea-ce-9276f4bbfb20dac51cdc0d675aef80b7f639a378/platform/core-api/src/com/intellij/openapi/project/Project.java), so `APlusProject#getCourseFilePath` would cause a null-pointer exception

Reply:
I've understood "default project" is some kind of placeholder object for "no project open" situation but I don't really know in which circumstances this actually occurs. Something we should research on at some point.